src/llvm-ptls.cpp

```diff
@@ -99,23 +100,41 @@ void LowerPTLS::runOnFunction(LLVMContext &ctx, Module &M, Function *F,
         ptlsStates->addAttribute(AttributeSet::FunctionIndex,
                                  Attribute::NoUnwind);
     }
-    else if (jl_tls_offset != -1) {
 #ifdef LLVM37
+    else if (jl_tls_offset != -1) {
         auto T_int8 = Type::getInt8Ty(ctx);
         auto T_pint8 = PointerType::get(T_int8, 0);
-        auto T_size = (sizeof(size_t) == 8 ? Type::getInt64Ty(ctx) :
-                       Type::getInt32Ty(ctx));
         // Replace the function call with inline assembly if we know
         // how to generate it.
-        const char *asm_str = nullptr;
+#  if defined(_CPU_X86_64_) || defined(_CPU_X86_)
+        // Workaround LLVM bug by hiding the offset computation
+        // (and therefore the optimization opportunity) from LLVM.
+        static const std::string asm_str = [&] () {
+            std::stringstream stm;
 #  if defined(_CPU_X86_64_)
-        asm_str = "movq %fs:0, $0";
-#  elif defined(_CPU_X86_)
-        asm_str = "movl %gs:0, $0";
-#  elif defined(_CPU_AARCH64_)
-        asm_str = "mrs $0, tpidr_el0";
+            stm << "movq %fs:0, $0;\naddq $$" << jl_tls_offset << ", $0";
+#  else
+            stm << "movl %gs:0, $0;\naddl $$" << jl_tls_offset << ", $0";
 #  endif
-        assert(asm_str && "Cannot emit thread pointer for this architecture.");
+            return stm.str();
+        }();
+        // The add instruction clobbers flags
+        auto tp = InlineAsm::get(FunctionType::get(T_pint8, false),
+                                 asm_str.c_str(),
+                                 "=r,~{dirflag},~{fpsr},~{flags}", false);
```

Comment:
Mainly because this is what clang generates and I couldn't find a single documentation of what each one mean and can't really figure out the logic in the x86 target code that uses these flags..